src/java.base/share/classes/java/util/stream/Gatherer.java

```diff
@@ -0,0 +1,571 @@
+ *     the integrator until it returns {@code false}, and then joining each
+ *     partitions state using the combiner, and then invoking the finisher on
+ *     the joined state. Outputs and state later in the input sequence will
+ *     be discarded if processing an earlier segment short-circuits.</li>
+ *     <li>Gatherers whose finisher is {@link #defaultFinisher()} are considered
+ *     to not have an end-of-stream hook and invoking their finisher is
+ *     optional.</li>
+ * </ul>
+ *
+ * @see Stream#gather(Gatherer)
+ * @see Gatherers
+ *
+ * @param <T> the type of input elements to the gatherer operation
+ * @param <A> the potentially mutable state type of the gatherer operation
+ *            (often hidden as an implementation detail)
+ * @param <R> the type of output elements from the gatherer operation
+ * @since 22
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.STREAM_GATHERERS)
+public interface Gatherer<T, A, R> {
+    /**
+     * A function that produces an instance of the intermediate state used for
+     * this gathering operation.
+     *
+     * @implSpec The implementation in this interface returns
+     *           {@link #defaultInitializer()}.
+     *
+     * @return A function that produces an instance of the intermediate state
+     * used for this gathering operation
+     */
+    default Supplier<A> initializer() {
+        return defaultInitializer();
+    };
+
+    /**
+     * A function which integrates provided elements, potentially using
+     * the provided intermediate state, optionally producing output to the
+     * provided {@link Downstream}.
+     *
+     * @return a function which integrates provided elements, potentially using
+     *         the provided state, optionally producing output to the provided
+     *         Downstream
+     */
+    Integrator<A, T, R> integrator();
+
+    /**
+     * A function which accepts two intermediate states and combines them into
+     * one.
+     *
+     * @implSpec The implementation in this interface returns
+     *           {@link #defaultCombiner()}.
+     *
+     * @return a function which accepts two intermediate states and combines
+     *         them into one
+     */
+    default BinaryOperator<A> combiner() {
+        return defaultCombiner();
+    }
+
+    /**
+     * A function which accepts the final intermediate state
+     * and a {@link Downstream} object, allowing to perform a final action at
+     * the end of input elements.
+     *
+     * @implSpec The implementation in this interface returns
+     *           {@link #defaultFinisher()}.
+     *
+     * @return a function which transforms the intermediate result to the final
+     *         result(s) which are then passed on to the provided Downstream
+     */
+    default BiConsumer<A, Downstream<? super R>> finisher() {
+        return defaultFinisher();
+    }
+
+    /**
+     * Returns a composed Gatherer which connects the output of this Gatherer
```

Comment:
Generally, it would if we could harmonize and consolidate return clauses using {@return ...}

Reply:
Yeah, not in this PR tho :)